### ISSUE TITLE
converting VO table to a fits file: raise ValueError('Illegal format %s.' % format) ValueError: Illegal format object.

### ISSUE THREAD

**Issue by richardgmcmahon** [astropy/astropy#5258]
I have a VO table from the Gaia archive that I want to convert to fits format. 

The vot format file is valid and has a column attribute format which have value %r. There is a mixture of 32 and 64 bit floating point numbers, bool, strings, ints.

When I try to write a fits file I get the error.

   raise ValueError('Illegal format %s.' % format)
ValueError: Illegal format object.

A simple fix would be to remove/delete  the vot column attribute format. Is this possible? If yes, how do I do it? It is not obvious to me from the docs.

Alternatively is there a valid way to convert prior to writing the fits file since this might be useful too although removal would meet my needs and that is what topcat seems to do.

I can look into the conversion based on TFORM is described here: https://heasarc.gsfc.nasa.gov/docs/software/fitsio/quick/node10.html

**Comment by pllim** [comment]
@richardgmcmahon , is it possible to provide a sample of the table in question? Doesn't have to be the whole table, just enough to reproduce the problem. Thanks.

**Comment by jslavin** [comment]
I'm having a similar issue.  In this case I used the ds9 catalog tool to compare two catalogs of sources.  When I saved it, the merged catalog was a VO table.  I then read it with astropy.io.votable.parse_single_table
which gave me a normal seeming table.  But when I tried to write it out as a fits file using table.write('source_match.fits'), I got 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-487e9212ab64> in <module>()
----> 1 table.write('match_HFC.fits', format='fits')

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/table/table.pyc in write(self, *args, **kwargs)
   2347         passed through to the underlying data reader (e.g. `~astropy.io.ascii.write`).
   2348         """
-> 2349         io_registry.write(self, *args, **kwargs)
   2350 
   2351     def copy(self, copy_data=True):

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/registry.pyc in write(data, *args, **kwargs)
    401 
    402     writer = get_writer(format, data.__class__)
--> 403     writer(data, *args, **kwargs)
    404 
    405 

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/fits/connect.pyc in write_table_fits(input, output, overwrite)
    208     """
    209 
--> 210     table_hdu = table_to_hdu(input)
    211 
    212     # Check if output file already exists

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/fits/convenience.pyc in table_to_hdu(table)
    457                 column.set_fill_value(np.nan)
    458 
--> 459         fits_rec = FITS_rec.from_columns(np.array(table.filled()))
    460         table_hdu = BinTableHDU(fits_rec)
    461         for col in table_hdu.columns:

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/fits/fitsrec.pyc in from_columns(cls, columns, nrows, fill)
    315 
    316         if not isinstance(columns, ColDefs):
--> 317             columns = ColDefs(columns)
    318 
    319         # read the delayed data

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.pyc in __init__(self, input, tbtype, ascii)
   1217         elif isinstance(input, np.ndarray) and input.dtype.fields is not None:
   1218             # Construct columns from the fields of a record array
-> 1219             self._init_from_array(input)
   1220         elif isiterable(input):
   1221             # if the input is a list of Columns

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.pyc in _init_from_array(self, array)
   1252             cname = array.dtype.names[idx]
   1253             ftype = array.dtype.fields[cname][0]
-> 1254             format = self._col_format_cls.from_recformat(ftype)
   1255 
   1256             # Determine the appropriate dimensions for items in the column

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.pyc in from_recformat(cls, recformat)
    230         """Creates a column format from a Numpy record dtype format."""
    231 
--> 232         return cls(_convert_format(recformat, reverse=True))
    233 
    234     @lazyproperty

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.pyc in _convert_format(format, reverse)
   2227 
   2228     if reverse:
-> 2229         return _convert_record2fits(format)
   2230     else:
   2231         return _convert_fits2record(format)

/export/slavin/python/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.pyc in _convert_record2fits(format)
   2190         output_format = repeat + NUMPY2FITS[recformat]
   2191     else:
-> 2192         raise ValueError('Illegal format %s.' % format)
   2193 
   2194     return output_format

ValueError: Illegal format object.

Ah, I see that the dtype of every column has been set to 'object' for some reason.  That seems to be the root of the problem.  How would I fix that?

**Comment by pllim** [comment]
@embray , any insight here? Thanks!

**Comment by pllim** [comment]
@jslavin , I get the same traceback with data provided by @richardgmcmahon . Indeed, I think it's the same problem.

**Comment by embray** [comment]
Somehow or another one of the tables contains a column of dtype=object

**Comment by pllim** [comment]
So, it's actually bad data and not a bug in the codes?

**Comment by richardgmcmahon** [comment]
In my vot table there is no dtype=object; i.e. data is OK; 

There is a mixture of floar32, float64, int16, int32, int64 and  bool.

It looks might be a problem with the conversion from VOT to  FITS;

Looking at my traceback the problem could be in: io/fits

It could be either a 'dtype' or a 'format' keyword issue.



My traceback is:

Traceback (most recent call last):
  File "rd_vot.py", line 43, in <module>
    table.write(outputFileName, format='fits')
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/table/table.py", line 2349, in write
    io_registry.write(self, *args, **kwargs)
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/registry.py", line 403, in write
    writer(data, *args, **kwargs)
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/fits/connect.py", line 210, in write_table_fits
    table_hdu = table_to_hdu(input)
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/fits/convenience.py", line 459, in table_to_hdu
    fits_rec = FITS_rec.from_columns(np.array(table.filled()))
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/fits/fitsrec.py", line 317, in from_columns
    columns = ColDefs(columns)
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.py", line 1219, in __init__
    self._init_from_array(input)
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.py", line 1254, in _init_from_array
    format = self._col_format_cls.from_recformat(ftype)
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.py", line 232, in from_recformat
    return cls(_convert_format(recformat, reverse=True))
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.py", line 2229, in _convert_format
    return _convert_record2fits(format)
  File "/home/rgm/anaconda/lib/python2.7/site-packages/astropy/io/fits/column.py", line 2192, in _convert_record2fits
    raise ValueError('Illegal format %s.' % format)
ValueError: Illegal format object.

**Comment by saimn** [comment]
@richardgmcmahon : you have a column that it read with the object dtype it seems:

```
In [8]: set(c.dtype for c in t.columns.values())
Out[8]: 
{dtype('bool'),
 dtype('int16'),
 dtype('float32'),
 dtype('int32'),
 dtype('int64'),
 dtype('float64'),
 dtype('O')}

In [14]: t['phot_variable_flag'].data.data
Out[14]: 
array([b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'',
       b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'',
       b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'',
       b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'',
       b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'',
       b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'', b'',
       b'', b'', b'', b'', b''], dtype=object)
```

**Comment by richardgmcmahon** [comment]
Ok, thanks for digging into this; I can reproduce @saimn finds.

Interesting topcat says thes column phot_variable_flag has datatype = char

Another flag column called: 'astrometric_primary_flag' is read by topcat and astropy as dtype=boolean.

I have reported the datatype problem to the Gaia archive.

Meanwhil I have deleted the column and Table.write can write a FITS file.

Then I tried a column replacement which seems to work:

colname = 'phot_variable_flag'
coldata = table[colname].data
NewColumn = Table.Column(coldata, dtype = 'bool')
table.replace_column(colname, NewColumn)

table.write('tmp.fits', overwrite=True)
